api/dbadapters/bigquery.ts

```diff
@@ -257,54 +248,82 @@ export class BigQueryDbAdapter implements IDbAdapter {
+    options: {
+      onCancel: OnCancel;
+    }
+  ): Promise<void> {
+    if (allActions.length === 0) {
       return;
     }
-    const allInvolvedTargets = new StringifiedSet(JSONObjectStringifier.create<dataform.ITarget>());
-    actions.forEach(action => {
-      allInvolvedTargets.add(action.target);
-      action.transitiveInputs.forEach(transitiveInput => allInvolvedTargets.add(transitiveInput));
-    });
-
-    const tableMetadataByTarget = new StringifiedMap<dataform.ITarget, dataform.ITableMetadata>(
-      JSONObjectStringifier.create()
-    );
-    await Promise.all(
-      Array.from(allInvolvedTargets).map(async target => {
-        tableMetadataByTarget.set(target, await this.table(target));
-      })
-    );
-    const queries = actions.map(action => {
-      const persistTable = dataform.PersistedTableMetadata.create({
-        target: action.target,
-        lastUpdatedMillis: tableMetadataByTarget.get(action.target).lastUpdatedMillis,
-        definitionHash: hashExecutionAction(action),
-        transitiveInputTables: action.transitiveInputs.map(transitiveInput => {
-          if (!tableMetadataByTarget.has(transitiveInput)) {
-            throw new Error(`Could not find table metadata for ${JSON.stringify(transitiveInput)}`);
-          }
-          return tableMetadataByTarget.get(transitiveInput);
-        })
-      });
-
-      const targetName = `${action.target.database}.${action.target.schema}.${action.target.name}`;
-
-      return buildQuery(targetName, persistTable);
-    });
+    const targetStringifier = JSONObjectStringifier.create<dataform.ITarget>();
+    try {
+      // Create the cache table, if needed.
+      await this.execute(
+        `
+CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+  target STRING,
+  metadata_proto STRING
+)`,
+        options
+      );
+      // Before saving any new data, delete all entries for 'allActions'.
+      await this.execute(
+        `
+DELETE \`${CACHED_STATE_TABLE_NAME}\` WHERE target IN (${allActions
+          .map(({ target }) => `'${targetStringifier.stringify(target)}'`)
+          .join(",")})`,
+        options
+      );
 
-    const unionQuery = queries.join(" UNION ALL ");
-
-    const updateQuery = `MERGE INTO \`${CACHED_STATE_TABLE_NAME}\` T
-    USING (${unionQuery}) S
-    ON (T.target_name = S.target_name)
-    WHEN NOT MATCHED THEN
-      INSERT (target_name, metadata_json, metadata_proto)
-      VALUES(S.target_name, S.metadata_json, S.metadata_proto)
-    WHEN MATCHED THEN
-      UPDATE SET metadata_json = S.metadata_json,
-          metadata_proto = S.metadata_proto;`;
-    await this.runQuery(updateQuery);
+      // Save entries for 'actionsToPersist'.
+      const transitiveInputMetadataByTarget = await this.computeTransitiveInputMetadataMap(
+        actionsToPersist
+      );
+      const valuesTuples = actionsToPersist
+        // If we were unable to load metadata for the action's output dataset, or for any of the action's
```

Comment:
this implies to me that the first time we create a table, we won't write a cache entry for it. This doesn't seem like what we want to happen?

Reply:
no, after the run completes we read all necessary metadata before we write it back. see `computeTransitiveInputMetadataMap`.